src/plugins/dashboard/public/dashboard_container/state/diffing/dashboard_diffing_functions.ts

```diff
@@ -52,6 +52,25 @@ export const isKeyEqual = async (
   return fastIsEqual(diffFunctionProps.currentValue, diffFunctionProps.lastValue);
 };
 
+export const isKeyEqual = (
+  key: keyof Omit<DashboardContainerInput, 'panels'>, // only Panels is async
+  diffFunctionProps: DiffFunctionProps<typeof key>,
+  diffingFunctions: DashboardDiffFunctions
+) => {
+  const propsAsNever = diffFunctionProps as never; // todo figure out why props has conflicting types in some constituents.
+  const diffingFunction = diffingFunctions[key];
+  if (!diffingFunction) {
+    return fastIsEqual(diffFunctionProps.currentValue, diffFunctionProps.lastValue);
+  }
+
+  if (diffingFunction?.prototype?.name === 'AsyncFunction') {
+    throw new Error(
+      `The function for key "${key}" is async, must use isKeyEqualAsync for asynchronous functions`
+    );
+  }
+  return diffingFunction(propsAsNever);
```

Comment:
**minor nit:** Hmm.... I wonder if we could make it clearer when we use `isKeyEqualAsync` versus `isKeyEqual` (synchronous) - either with a comment or perhaps by making the naming clearer? Like, `isKeyEqualAsync -> unsavedChangesIsKeyEqual` and `isKeyEqual -> shouldRefreshIsKeyEqual`... Not sure if that's too specific